### PR TITLE
Update getcommandpath to use the PATH variable set through setenv=PATH=...

### LINKED ISSUE #1146: DOUBT: Regarding PATH environment variable
How to make a command of `tox` modify the `PATH` environment variable, so that the other commands can use the modified `PATH` environment variable. As of now, if a command changes the `PATH` variable, then the changed `PATH` variable is not being used for the following commands.

Roughly, it is something like the following example
```
[testenv]
setenv = 
  PATH={env:PATH}:{env:HOME/install/bin}
passenv =
  PATH
commands =
  bash install_tool.sh
  tool
```
The command `bash install_tool.sh` would install and store a `binary executable` in `$HOME/install/bin`. Hence I have set the PATH initially, but finally the command `tool` is not working. The error I am getting is `InvocationError: could not find executable 'tool'`.

Could someone help me with this?

### PR DESCRIPTION
**Pull request by Bhargavasomu**
Fixes #1146

### CODE REVIEW

**Comment by Bhargavasomu** [comment]
Could you please help me with the errors that I am seeing.

File: src/tox/venv.py [hunk 1]
```diff
@@ -515,6 +515,11 @@ def _pcall(
         env = self._get_os_environ(is_test_command=is_test_command)
         bin_dir = str(self.envconfig.envbindir)
         env["PATH"] = os.pathsep.join([bin_dir, os.environ["PATH"]])
+        if "PATH" in self.envconfig.setenv:
+            env["PATH"] = os.pathsep.join([env["PATH"], self.envconfig.setenv["PATH"]])
+            # Remove redundantly duplicated directories in PATH
+            env["PATH"] = ':'.join(list(set(env["PATH"].split(":"))))
```
Comment (obestwalter) [inline src/tox/venv.py]: I don't think we should do that. A set is unordered and `PATH` is ordered - if I put myself in the shoes of the user, I would be surprised if that order would be arbitrarily changed. So I would completely remove that line.

**Comment by obestwalter** [comment]
https://toxdev.visualstudio.com/tox/_build/results?buildId=1152 - black needed to reformat your changes. Please run `tox -e fix_lint` locally.

**Comment by obestwalter** [comment]
The other error seems unrelated to your changes on first look but does not occur on master, so this needs looking into more closely. Please run the tests locally and have a closer look.

File: src/tox/venv.py [hunk 2]
```diff
@@ -515,6 +515,11 @@ def _pcall(
         env = self._get_os_environ(is_test_command=is_test_command)
         bin_dir = str(self.envconfig.envbindir)
         env["PATH"] = os.pathsep.join([bin_dir, os.environ["PATH"]])
+        if "PATH" in self.envconfig.setenv:
+            env["PATH"] = os.pathsep.join([env["PATH"], self.envconfig.setenv["PATH"]])
+            # Remove redundantly duplicated directories in PATH
```
Comment (gaborbernat) [inline src/tox/venv.py]: I like the idea, but you should make sure you don't re-order path values (use a set and a list, to keep the first appearance) - also this logic should run always not just when setenv is set

File: src/tox/venv.py [hunk 3]
```diff
@@ -515,6 +515,11 @@ def _pcall(
         env = self._get_os_environ(is_test_command=is_test_command)
         bin_dir = str(self.envconfig.envbindir)
         env["PATH"] = os.pathsep.join([bin_dir, os.environ["PATH"]])
+        if "PATH" in self.envconfig.setenv:
+            env["PATH"] = os.pathsep.join([env["PATH"], self.envconfig.setenv["PATH"]])
```
Comment (gaborbernat) [inline src/tox/venv.py]: this needs more thought, why join here? this so to say overrides the setenv 🤔 we should only join if the setenv includes the previous, not?

File: src/tox/venv.py [hunk 4]
```diff
@@ -515,6 +515,11 @@ def _pcall(
...
+            # Remove redundantly duplicated directories in PATH
+            env["PATH"] = ':'.join(list(set(env["PATH"].split(":"))))
+        os.environ["PATH"] = env["PATH"]
```
Comment (gaborbernat) [inline src/tox/venv.py]: we should never modify the global path, chaning the env should be all you need to propagate to the subprocess call

Comment (Bhargavasomu) [reply src/tox/venv.py]: @gaborbernat if we don't modify the global path, then while doing the `_normal_lookup`, it won't identify the executable. Attaching the snippet for reference
https://github.com/tox-dev/tox/blob/d6cd8d42fa06cc02d747ebe92b3179bc677f0591/src/tox/venv.py#L167-L168

Comment (gaborbernat) [reply src/tox/venv.py]: that's a bug for the normal lookup though, we should inject the environments generated paths https://py.readthedocs.io/en/latest/path.html#py._path.local.LocalPath.sysfind to fix it

File: src/tox/venv.py (hunk 3 above)
Comment (Bhargavasomu) [reply src/tox/venv.py]: Right, will have to do something about that.

**Comment by Bhargavasomu** [comment]
@gaborbernat @obestwalter how come `Tox CI` hasn't run? Also please review the recent changes, I have fixed this in kind of an elegant way to read and understand. Please let me know for further changes.

**Comment by Bhargavasomu** [comment]
Also when I installed this version of `tox`, the bug which I was personally experiencing got fixed !!

**Comment by Bhargavasomu** [comment]
On `python3.5` running the test `tests/unit/test_venv.py` is behaving randomly. Please apply this PR on a `python3.5` environment and see the test behaving randomly on multiple runs. Could someone please explain me why that is happening?

**Comment by Bhargavasomu** [comment]
EDIT : Fixed all the local errors in `py3.5` and `py3.4` etc. But it is not visible, because `Tox CI` is not running. Can someone please look into this?

**Comment by gaborbernat** [comment]
This PR still lacks tests, without it it's small chance of getting accepted. Please  add  unit test for each feature improvement one by one, taking care of edge cases too. Furthermore I would prefer to have a one place path mangling, now you do it in two places which makes the code hard to maintain.

**Comment by Bhargavasomu** [comment]
> path mangling, now you do it in two places which makes the code hard to maintain

@gaborbernat I understand your view, but the following issue was what I faced when I tried to `PATH` mangling only once.

In the following snippet, the path order to be maintained is `envbindir + os.environ["PATH"]`
https://github.com/tox-dev/tox/blob/d6cd8d42fa06cc02d747ebe92b3179bc677f0591/src/tox/venv.py#L514-L518.
But when I try to pass that variable to `_normal_lookup`, even `envbindir` would be passed, which in practice/theory is not correct. Hence I put it this way. Please correct me if I am wrong.

I will add test cases, in a while. Could you please let me know why `Tox CI` is not running?

**Comment by gaborbernat** [comment]
https://dev.azure.com/toxdev/tox/_build/results?buildId=1153

**Comment by Bhargavasomu** [comment]
@gaborbernat that was an old build. I have made 2 pushes after that. (force pushes). But it is not updating.

File: src/tox/venv.py [hunk 5]
```diff
@@ -514,7 +518,12 @@ def _pcall(
         os.environ.pop("VIRTUALENV_PYTHON", None)
         env = self._get_os_environ(is_test_command=is_test_command)
         bin_dir = str(self.envconfig.envbindir)
-        env["PATH"] = os.pathsep.join([bin_dir, os.environ["PATH"]])
+        if "PATH" in self.envconfig.setenv:
+            env["PATH"] = os.pathsep.join([bin_dir, self.envconfig.setenv["PATH"]])
+        else:
+            env["PATH"] = os.pathsep.join([bin_dir, os.environ["PATH"]])
+        # Remove redundantly duplicated directories in PATH without changing order
+        env["PATH"] = ":".join(list(dict.fromkeys(env["PATH"].split(":"))))
```
Comment (asottile) [inline src/tox/venv.py]: this is (for instance) incorrect on windows -- use `os.pathsep` instead of `':'` -- also it's unnecessary (and imo a surprising behaviour) to remove duplicates so this modification probably shouldn't happen at all

Comment (asottile) [reply src/tox/venv.py]: (above is true for `.split(":")` as well -- quite surprised this patch passes tests on windows?

Comment (Bhargavasomu) [reply src/tox/venv.py]: This is not passing on the windows test cases, and you gave me a solution. Thankyou.

**Comment by Bhargavasomu** [comment]
@asottile I have added the new testcase. Please check it out.

**Comment by gaborbernat** [comment]
I'll be honest I still don't entirely see the reason for this. What you describe sounds like you want to side step the tox as entry point which I find very dangerous. The more I think about it the surer I am that the only reasonable thing which we could allow is injecting elements into the path before the environment path, but after the tox environments bin folder.

**Comment by gaborbernat** [comment]
On the subject of simplifying the path I think it's ok and will help decrease log size and debug. We should definitely keep only the first instance of path occurances. We could also remove not existing paths (I'm not entirely settled on this yet though). That being said this change is orthogonal to this PR so maybe let's keep it separate.

**Comment by Bhargavasomu** [comment]
@asottile @gaborbernat I'm really confused from all the above comments. Could you please frame it in a better way as to do what next?

**Comment by gaborbernat** [comment]
@asottile was alluding to the fact that's probably more robust to alter the PATH from within the tests (via some session fixture e.g.) rather than inside tox. This way your tests will work also for cases when users are not calling the test from within tox (e.g. they do from within an IDE).

**Comment by Bhargavasomu** [comment]
Thankyou for the support.
